File: MyMusic.CLI/Services/Sync/SyncActionsDevice.cs
namespace MyMusic.CLI.Services.Sync;

using System.IO.Abstractions;
using System.Text.Json;
using Microsoft.Extensions.Logging;
using MyMusic.CLI.Services.Sync.Types;

public class SyncActionsDevice(
    IFileOps fileOps,
    ISyncApiClient apiClient,
    IUserPrompt userPrompt,
    IFileSystem fileSystem,
    ILogger<SyncActionsDevice> logger)
{
    public record ActionResult(
        string Action,
        string FilePath,
        string Source = "Device",
        string? Reason = null,
        string? ErrorMessage = null,
        long? SongId = null,
        long? RecordId = null,
        SyncActionCounts? Counts = null);

    public async Task<ActionResult> ActionCreateRemoteAsync(
        long deviceId,
        long sessionId,
        string repositoryPath,
        SyncFileInfo fileInfo,
        CancellationToken ct = default)
    {
        var fullPath = Path.Combine(repositoryPath, fileInfo.Path);
        if (!fileSystem.File.Exists(fullPath))
        {
            logger.LogWarning("File not found: {Path}", fullPath);
            return new ActionResult("Error", fileInfo.Path, Reason: "File not found");
        }

        try
        {
            await using var stream = fileSystem.File.OpenRead(fullPath);
            var fileName = Path.GetFileName(fullPath);
            var uploadResult = await apiClient.UploadFileAsync(deviceId, sessionId, new UploadFileRequest
            {
                FileStream = stream,
                FileName = fileName,
                Path = fileInfo.Path,
                ModifiedAt = fileInfo.ModifiedAt.ToUniversalTime().ToString("O"),
                CreatedAt = fileInfo.CreatedAt.ToUniversalTime().ToString("O")
            }, ct);

            return new ActionResult("Created", fileInfo.Path, Reason: fileInfo.Reason, SongId: uploadResult.SongId, Counts: uploadResult.Counts);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Failed to upload file: {Path}", fileInfo.Path);
            return new ActionResult("Error", fileInfo.Path, ErrorMessage: ex.Message, Reason: fileInfo.Reason);
        }
    }

    public async Task<ActionResult> ActionUpdateRemoteAsync(
        long deviceId,
        long sessionId,
        string repositoryPath,
        SyncFileInfo fileInfo,
        CancellationToken ct = default)
    {
        var fullPath = Path.Combine(repositoryPath, fileInfo.Path);
        if (!fileSystem.File.Exists(fullPath))
        {
            logger.LogWarning("File not found: {Path}", fullPath);
            return new ActionResult("Error", fileInfo.Path, Reason: "File not found");
        }

        try
        {
            await using var stream = fileSystem.File.OpenRead(fullPath);
            var fileName = Path.GetFileName(fullPath);
            var uploadResult = await apiClient.UploadFileAsync(deviceId, sessionId, new UploadFileRequest
            {
                FileStream = stream,
                FileName = fileName,
                Path = fileInfo.Path,
                ModifiedAt = fileInfo.ModifiedAt.ToUniversalTime().ToString("O"),
                CreatedAt = fileInfo.CreatedAt.ToUniversalTime().ToString("O")
            }, ct);

            return new ActionResult("Updated", fileInfo.Path, Reason: fileInfo.Reason, SongId: uploadResult.SongId, Counts: uploadResult.Counts);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Failed to update file: {Path}", fileInfo.Path);
            return new ActionResult("Error", fileInfo.Path, ErrorMessage: ex.Message, Reason: fileInfo.Reason);
        }
    }

    public async Task<ActionResult?> ActionCreateLocalAsync(
        long deviceId,
        long sessionId,
        string repositoryPath,
        long? songId,
        string relativePath,
        bool dryRun,
        bool autoConfirm,
        long recordId,
        string? reason = null,
        CancellationToken ct = default)
    {
        var fullPath = Path.Combine(repositoryPath, relativePath);
        var fileExists = fileOps.FileExists(fullPath);

        if (fileExists)
        {
            logger.LogError("File already exists during create: {Path}", relativePath);
            return new ActionResult("Error", relativePath, Source: "Server", ErrorMessage: "File already exists", Reason: "Unexpected local file during create");
        }

        return await DownloadAndAckAsync(deviceId, sessionId, repositoryPath, songId, relativePath, dryRun, recordId, reason, isUpdate: false, ct);
    }

    public async Task<ActionResult?> ActionUpdateLocalAsync(
        long deviceId,
        long sessionId,
        string repositoryPath,
        long? songId,
        string relativePath,
        bool dryRun,
        bool autoConfirm,
        long recordId,
        string? reason = null,
        CancellationToken ct = default)
    {
        var fullPath = Path.Combine(repositoryPath, relativePath);
        var fileExists = fileOps.FileExists(fullPath);

        if (!fileExists)
        {
            logger.LogError("File not found during update: {Path}", relativePath);
            return new ActionResult("Error", relativePath, Source: "Server", ErrorMessage: "File not found", Reason: "Missing local file during update");
        }

        return await DownloadAndAckAsync(deviceId, sessionId, repositoryPath, songId, relativePath, dryRun, recordId, reason, isUpdate: true, ct);
    }

    private async Task<ActionResult?> DownloadAndAckAsync(
        long deviceId,
        long sessionId,
        string repositoryPath,
        long? songId,
        string relativePath,
        bool dryRun,
        long recordId,
        string? reason,
        bool isUpdate,
        CancellationToken ct)
    {
        var actionName = isUpdate ? "UpdateLocal" : "CreateLocal";
        var baseReason = reason ?? (isUpdate ? "Server-initiated update" : "Server-initiated download");
        var fullPath = Path.Combine(repositoryPath, relativePath);

        if (dryRun)
        {
            var ackResult = await apiClient.AcknowledgeActionAsync(deviceId, sessionId, new AcknowledgeActionRequest
            {
                RecordIds = [recordId]
            }, ct);

            return new ActionResult(
                actionName,
                relativePath,
                Source: "Server",
                Reason: baseReason,
                SongId: songId,
                RecordId: recordId,
                Counts: ackResult.Counts);
        }

        var tempPath = fullPath + ".tmp";
        try
        {
            await fileOps.EnsureDirectoryAsync(fullPath, ct);

            await using var stream = await apiClient.DownloadSongAsync(songId!.Value, ct);
            await fileOps.WriteFileAsync(tempPath, stream, ct);

            if (isUpdate)
            {
                await fileOps.DeleteFileAsync(fullPath, ct);
            }

            await fileOps.MoveFileAsync(tempPath, fullPath, ct);

            var modifiedAt = await fileOps.GetModificationTimeAsync(fullPath, ct);
            var ackResult = await apiClient.AcknowledgeActionAsync(deviceId, sessionId, new AcknowledgeActionRequest
            {
                RecordIds = [recordId],
                ModifiedAt = modifiedAt
            }, ct);

            return new ActionResult(
                actionName,
                relativePath,
                Source: "Server",
                Reason: baseReason,
                SongId: songId,
                RecordId: recordId,
                Counts: ackResult.Counts);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Failed to {Action} file: {Path}", isUpdate ? "update" : "download", relativePath);
            return new ActionResult("Error", relativePath, Source: "Server", ErrorMessage: ex.Message, Reason: $"{baseReason} failed", SongId: songId);
        }
        finally
        {
            if (fileOps.FileExists(tempPath))
            {
                await fileOps.DeleteFileAsync(tempPath, ct);
            }
        }
    }

    public async Task<ActionResult?> ActionDeleteAsync(
        long deviceId,
        long sessionId,
        string repositoryPath,
        long? songId,
        string relativePath,
        bool dryRun,
        bool autoConfirm,
        long recordId,
        string? reason = null,
        CancellationToken ct = default)
    {
        var fullPath = Path.Combine(repositoryPath, relativePath);
        var fileExists = fileOps.FileExists(fullPath);

        if (!fileExists)
        {
            await apiClient.AcknowledgeActionAsync(deviceId, sessionId, new AcknowledgeActionRequest
            {
                RecordIds = [recordId]
            }, ct);
            return null;
        }

        if (!dryRun && !autoConfirm)
        {
            var confirmed = await userPrompt.ConfirmDeletionAsync(relativePath, ct);
            if (!confirmed)
            {
                return null;
            }
        }

        var baseReason = reason ?? "Server-initiated removal";

        if (dryRun)
        {
            var ackResult = await apiClient.AcknowledgeActionAsync(deviceId, sessionId, new AcknowledgeActionRequest
            {
                RecordIds = [recordId]
            }, ct);

            return new ActionResult("Delete", relativePath, Source: "Server", Reason: baseReason, SongId: songId, RecordId: recordId, Counts: ackResult.Counts);
        }

        try
        {
            await fileOps.DeleteFileAsync(fullPath, ct);
            var ackResult = await apiClient.AcknowledgeActionAsync(deviceId, sessionId, new AcknowledgeActionRequest
            {
                RecordIds = [recordId]
            }, ct);

            return new ActionResult("Delete", relativePath, Source: "Server", Reason: baseReason, SongId: songId, RecordId: recordId, Counts: ackResult.Counts);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Failed to delete file: {Path}", relativePath);
            return new ActionResult("Error", relativePath, Source: "Server", ErrorMessage: ex.Message, Reason: $"{baseReason} failed", SongId: songId);
        }
    }

    public async Task<ActionResult?> ActionRenameAsync(
        long deviceId,
        long sessionId,
        string repositoryPath,
        string relativePath,
        string previousRelativePath,
        bool dryRun,
        long recordId,
        CancellationToken ct = default)
    {
        var fullPath = Path.Combine(repositoryPath, relativePath);
        var previousFullPath = Path.Combine(repositoryPath, previousRelativePath);

        if (dryRun)
        {
            var ackResult = await apiClient.AcknowledgeActionAsync(deviceId, sessionId, new AcknowledgeActionRequest
            {
                RecordIds = [recordId]
            }, ct);

            return new ActionResult("Renamed", relativePath, Source: "Server", Reason: $"Renamed from '{previousRelativePath}'", RecordId: recordId, Counts: ackResult.Counts);
        }

        try
        {
            if (fileOps.FileExists(previousFullPath))
            {
                await fileOps.EnsureDirectoryAsync(fullPath, ct);
                await fileOps.MoveFileAsync(previousFullPath, fullPath, ct);
                fileOps.CleanupEmptyParentDirectories(previousFullPath, repositoryPath);
            }

            var ackResult = await apiClient.AcknowledgeActionAsync(deviceId, sessionId, new AcknowledgeActionRequest
            {
                RecordIds = [recordId]
            }, ct);

            return new ActionResult("Renamed", relativePath, Source: "Server", Reason: $"Renamed from '{previousRelativePath}'", RecordId: recordId, Counts: ackResult.Counts);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Failed to rename file: {PreviousPath} -> {Path}", previousRelativePath, relativePath);
            return new ActionResult("Error", relativePath, Source: "Server", ErrorMessage: ex.Message, Reason: $"Rename from '{previousRelativePath}' failed");
        }
    }

    public async Task<ResolveConflictsActionResult> ActionConflictAsync(
        long deviceId,
        long sessionId,
        string repositoryPath,
        List<SyncRecordItem> conflictRecords,
        List<SyncRecordItem> updateLocalRecords,
        CancellationToken ct = default)
    {
        var resolveItems = new List<ConflictResolveItem>();
        foreach (var conflict in conflictRecords)
        {
            if (!conflict.SongId.HasValue)
            {
                logger.LogWarning("Skipping conflict with no SongId: {Path}", conflict.FilePath);
                continue;
            }

            var fullPath = Path.Combine(repositoryPath, conflict.FilePath);
            if (!fileSystem.File.Exists(fullPath))
            {
                logger.LogWarning("Conflict file not found locally: {Path}", conflict.FilePath);
                continue;
            }

            var fileContentBase64 = await fileOps.ReadFileBase64Async(fullPath, ct);
            var conflictData = SyncDataDeserialization.DeserializeConflictCheckData(conflict.Data);
            resolveItems.Add(new ConflictResolveItem
            {
                Path = conflict.FilePath,
                SongId = conflict.SongId.Value,
                FileContentBase64 = fileContentBase64,
                LocalModifiedAt = conflictData?.LocalModifiedAt ?? DateTime.UtcNow
            });
        }

        var potentialUpdateItems = new List<PotentialUpdateResolveItem>();
        foreach (var update in updateLocalRecords)
        {
            if (!update.SongId.HasValue)
            {
                logger.LogWarning("Skipping potential update with no SongId: {Path}", update.FilePath);
                continue;
            }

            var fullPath = Path.Combine(repositoryPath, update.FilePath);
            if (!fileSystem.File.Exists(fullPath))
            {
                logger.LogWarning("Potential update file not found locally: {Path}", update.FilePath);
                continue;
            }

            var fileContentBase64 = await fileOps.ReadFileBase64Async(fullPath, ct);
            var updateData = SyncDataDeserialization.DeserializeUpdateLocalCheckData(update.Data);
            potentialUpdateItems.Add(new PotentialUpdateResolveItem
            {
                Path = update.FilePath,
                SongId = update.SongId.Value,
                FileContentBase64 = fileContentBase64,
                LocalModifiedAt = updateData?.LocalModifiedAt ?? DateTime.UtcNow,
                LastSyncedAt = updateData?.LastSyncedAt ?? DateTime.UtcNow
            });
        }

        if (resolveItems.Count == 0 && potentialUpdateItems.Count == 0)
        {
            return new ResolveConflictsActionResult(Records: [], Counts: SyncActionCounts.Empty);
        }

        try
        {
            var resolveResponse = await apiClient.ResolveConflictsAsync(deviceId, sessionId, new ResolveConflictsRequest
            {
                Conflicts = resolveItems,
                PotentialUpdates = potentialUpdateItems
            }, ct);

            foreach (var record in resolveResponse.Records)
            {
                var logMessage = record.Action switch
                {
                    SyncRecordAction.Conflict => $"Conflict for {record.FilePath}: {record.Reason}",
                    SyncRecordAction.UpdateTimestamp => $"Resolved conflict for {record.FilePath}: {record.Reason}",
                    SyncRecordAction.UpdateLocal => $"Created UpdateLocal action for record {record.Id}",
                    SyncRecordAction.Rename => $"Created Rename action for record {record.Id}",
                    SyncRecordAction.Error => $"Error for {record.FilePath}: {record.Reason}",
                    _ => $"Created {record.Action} action for record {record.Id}"
                };
                logger.LogInformation("{LogMessage}", logMessage);
            }

            return new ResolveConflictsActionResult(
                Records: resolveResponse.Records,
                Counts: resolveResponse.Counts);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Failed to resolve conflicts");
            return new ResolveConflictsActionResult(Records: [], Counts: SyncActionCounts.Empty);
        }
    }
}

public record ResolveConflictsActionResult(List<SyncRecordItem> Records, SyncActionCounts Counts);

file record ConflictCheckData(DateTime LocalModifiedAt, DateTime ServerModifiedAt);
file record UpdateLocalCheckData(DateTime LocalModifiedAt, DateTime ServerModifiedAt, DateTime LastSyncedAt);

file static class SyncDataDeserialization
{
    private static readonly JsonSerializerOptions Options = new() { PropertyNameCaseInsensitive = true };

    internal static ConflictCheckData? DeserializeConflictCheckData(System.Text.Json.JsonElement? data)
    {
        if (!data.HasValue || data.Value.ValueKind == System.Text.Json.JsonValueKind.Null) return null;
        try { return JsonSerializer.Deserialize<ConflictCheckData>(data.Value, Options); }
        catch { return null; }
    }

    internal static UpdateLocalCheckData? DeserializeUpdateLocalCheckData(System.Text.Json.JsonElement? data)
    {
        if (!data.HasValue || data.Value.ValueKind == System.Text.Json.JsonValueKind.Null) return null;
        try { return JsonSerializer.Deserialize<UpdateLocalCheckData>(data.Value, Options); }
        catch { return null; }
    }
}

File: MyMusic.Mobile/src/services/sync/sync-actions-device.ts
import type {ISyncApiClient, IFileOps, IUserPrompt, SyncContext, SyncFileBase, ActionResult, ResolveConflictsResult, SyncActionCounts, ProgressHandler, SyncRecordItem} from './types';
import type {SyncConflictResolveItem, SyncPotentialUpdateResolveItem, RenameData} from '../../api/types';
import {addDeltaToResult} from './types';
import {safeToIsoString} from './utils';

const EMPTY_COUNTS: SyncActionCounts = {
    createRemoteCount: 0,
    updateRemoteCount: 0,
    skippedCount: 0,
    createLocalCount: 0,
    updateLocalCount: 0,
    deleteCount: 0,
    linkCount: 0,
    unlinkCount: 0,
    renameCount: 0,
    conflictCount: 0,
    updateTimestampCount: 0,
    errorCount: 0,
};

export async function actionCreateRemote(
    apiClient: ISyncApiClient,
    fileOps: IFileOps,
    ctx: SyncContext,
    file: SyncFileBase,
    reason?: string
): Promise<ActionResult> {
    if (!fileOps.fileExists(file.fullPath)) {
        return {
            action: 'Error',
            filePath: file.relativePath,
            source: 'Device',
            reason,
            errorMessage: `File not found: ${file.fullPath}`,
        };
    }

    try {
        const filename = file.relativePath.split('/').pop();
        if (!filename) {
            throw new Error(`Invalid file path: relativePath is empty for ${file.fullPath}`);
        }

        const uploadResult = await apiClient.uploadFile(
            ctx.deviceId,
            ctx.sessionId!,
            {
                uri: file.fullPath,
                name: filename,
            },
            file.relativePath,
            safeToIsoString(file.modifiedAt)!,
            safeToIsoString(file.createdAt)!
        );

        return {
            action: 'CreateRemote',
            filePath: file.relativePath,
            source: 'Device',
            reason,
            songId: uploadResult.songId ?? undefined,
            counts: uploadResult.counts,
        };
    } catch (e) {
        const errorMessage = e instanceof Error ? e.message : String(e);
        return {
            action: 'Error',
            filePath: file.relativePath,
            source: 'Device',
            reason,
            errorMessage,
        };
    }
}

export async function actionUpdateRemote(
    apiClient: ISyncApiClient,
    fileOps: IFileOps,
    ctx: SyncContext,
    file: SyncFileBase,
    reason?: string
): Promise<ActionResult> {
    if (!fileOps.fileExists(file.fullPath)) {
        return {
            action: 'Error',
            filePath: file.relativePath,
            source: 'Device',
            reason,
            errorMessage: `File not found: ${file.fullPath}`,
        };
    }

    try {
        const filename = file.relativePath.split('/').pop();
        if (!filename) {
            throw new Error(`Invalid file path: relativePath is empty for ${file.fullPath}`);
        }

        const uploadResult = await apiClient.uploadFile(
            ctx.deviceId,
            ctx.sessionId!,
            {
                uri: file.fullPath,
                name: filename,
            },
            file.relativePath,
            safeToIsoString(file.modifiedAt)!,
            safeToIsoString(file.createdAt)!
        );

        return {
            action: 'UpdateRemote',
            filePath: file.relativePath,
            source: 'Device',
            reason,
            songId: uploadResult.songId ?? undefined,
            counts: uploadResult.counts,
        };
    } catch (e) {
        const errorMessage = e instanceof Error ? e.message : String(e);
        return {
            action: 'Error',
            filePath: file.relativePath,
            source: 'Device',
            reason,
            errorMessage,
        };
    }
}

export async function actionCreateLocal(
    apiClient: ISyncApiClient,
    fileOps: IFileOps,
    userPrompt: IUserPrompt,
    ctx: SyncContext,
    songId: number | null,
    path: string,
    decodedRepoPath: string,
    recordId: number,
    reason?: string
): Promise<ActionResult | null> {
    const fullPath = `${decodedRepoPath}/${path}`;
    const fileExists = fileOps.fileExists(fullPath);

    if (!ctx.options.dryRun && fileExists && !ctx.options.autoConfirm) {
        const confirmed = await userPrompt.confirmDeletion(path);
        if (!confirmed) {
            return null;
        }
    }

    const baseReason = reason ?? 'Server-initiated download';

    if (ctx.options.dryRun) {
        const ackResult = await apiClient.acknowledgeAction(ctx.deviceId, ctx.sessionId!, {
            recordIds: [recordId],
        });
        return {
            action: 'CreateLocal',
            filePath: path,
            source: 'Server',
            reason: baseReason,
            songId: songId ?? undefined,
            recordId,
            counts: ackResult.counts,
        };
    }

    try {
        await fileOps.ensureDirectory(fullPath);

        const tempPath = `${fullPath}.tmp`;
        try {
            const blob = await apiClient.downloadSong(songId!);
            await fileOps.writeFile(tempPath, blob);

            if (fileExists) {
                await fileOps.deleteFile(fullPath);
            }

            await fileOps.moveFile(tempPath, fullPath);
        } finally {
            if (fileOps.fileExists(tempPath)) {
                await fileOps.deleteFile(tempPath);
            }
        }

        const modifiedAt = fileOps.getModificationTime(fullPath);
        const ackResult = await apiClient.acknowledgeAction(ctx.deviceId, ctx.sessionId!, {
            recordIds: [recordId],
            modifiedAt: modifiedAt ? safeToIsoString(modifiedAt) : undefined,
        });

        return {
            action: 'CreateLocal',
            filePath: path,
            source: 'Server',
            reason: baseReason,
            songId: songId ?? undefined,
            recordId,
            counts: ackResult.counts,
        };
    } catch (e) {
        const errorMessage = e instanceof Error ? e.message : String(e);
        return {
            action: 'Error',
            filePath: path,
            source: 'Server',
            reason: `${baseReason} failed`,
            errorMessage,
            songId: songId ?? undefined,
        };
    }
}

export async function actionUpdateLocal(
    apiClient: ISyncApiClient,
    fileOps: IFileOps,
    userPrompt: IUserPrompt,
    ctx: SyncContext,
    songId: number | null,
    path: string,
    decodedRepoPath: string,
    recordId: number,
    reason?: string
): Promise<ActionResult | null> {
    return actionCreateLocal(apiClient, fileOps, userPrompt, ctx, songId, path, decodedRepoPath, recordId, reason);
}

export async function actionDelete(
    apiClient: ISyncApiClient,
    fileOps: IFileOps,
    userPrompt: IUserPrompt,
    ctx: SyncContext,
    path: string,
    decodedRepoPath: string,
    songId?: number,
    recordId?: number,
    reason?: string
): Promise<ActionResult | null> {
    const fullPath = `${decodedRepoPath}/${path}`;
    const fileExists = fileOps.fileExists(fullPath);

    if (!fileExists) {
        if (recordId) {
            const ackResult = await apiClient.acknowledgeAction(ctx.deviceId, ctx.sessionId!, { recordIds: [recordId] });
            ctx.result = addDeltaToResult(ctx.result, ackResult.counts ?? EMPTY_COUNTS);
        }
        return null;
    }

    let shouldDelete = true;

    if (!ctx.options.autoConfirm && !ctx.options.dryRun) {
        shouldDelete = await userPrompt.confirmDeletion(path);
    }

    if (!shouldDelete) {
        return null;
    }

    const baseReason = reason ?? 'Server-initiated removal';

    if (ctx.options.dryRun) {
        const ackResult = await apiClient.acknowledgeAction(ctx.deviceId, ctx.sessionId!, { recordIds: [recordId ?? 0] });
        return {
            action: 'Delete',
            filePath: path,
            source: 'Server',
            reason: baseReason,
            songId,
            recordId,
            counts: ackResult.counts,
        };
    }

    try {
        await fileOps.deleteFile(fullPath);
        const ackResult = await apiClient.acknowledgeAction(ctx.deviceId, ctx.sessionId!, { recordIds: [recordId ?? 0] });

        return {
            action: 'Delete',
            filePath: path,
            source: 'Server',
            reason: baseReason,
            songId,
            recordId,
            counts: ackResult.counts,
        };
    } catch (e) {
        const errorMessage = e instanceof Error ? e.message : String(e);
        return {
            action: 'Error',
            filePath: path,
            source: 'Server',
            reason: `${baseReason} failed`,
            errorMessage,
            songId,
        };
    }
}

export async function actionRename(
    apiClient: ISyncApiClient,
    fileOps: IFileOps,
    ctx: SyncContext,
    relativePath: string,
    previousRelativePath: string,
    decodedRepoPath: string,
    recordId: number
): Promise<ActionResult> {
    const fullPath = `${decodedRepoPath}/${relativePath}`;
    const previousFullPath = `${decodedRepoPath}/${previousRelativePath}`;

    if (ctx.options.dryRun) {
        const ackResult = await apiClient.acknowledgeAction(ctx.deviceId, ctx.sessionId!, {
            recordIds: [recordId],
        });
        return {
            action: 'Rename',
            filePath: relativePath,
            source: 'Server',
            reason: `Renamed from '${previousRelativePath}'`,
            recordId,
            counts: ackResult.counts,
        };
    }

    try {
        if (fileOps.fileExists(previousFullPath)) {
            await fileOps.ensureDirectory(fullPath);
            await fileOps.moveFile(previousFullPath, fullPath);
            await fileOps.deleteEmptyDirectories(previousFullPath, decodedRepoPath);
        }

        const ackResult = await apiClient.acknowledgeAction(ctx.deviceId, ctx.sessionId!, {
            recordIds: [recordId],
        });

        return {
            action: 'Rename',
            filePath: relativePath,
            source: 'Server',
            reason: `Renamed from '${previousRelativePath}'`,
            recordId,
            counts: ackResult.counts,
        };
    } catch (e) {
        const errorMessage = e instanceof Error ? e.message : String(e);
        return {
            action: 'Error',
            filePath: relativePath,
            source: 'Server',
            reason: `Rename from '${previousRelativePath}' failed`,
            errorMessage,
        };
    }
}

export async function actionConflict(
    apiClient: ISyncApiClient,
    fileOps: IFileOps,
    userPrompt: IUserPrompt,
    ctx: SyncContext,
    conflictRecords: SyncRecordItem[],
    updateLocalRecords: SyncRecordItem[],
    toUpdatePaths: Set<string>,
    onProgress: ProgressHandler
): Promise<ResolveConflictsResult> {
    if (conflictRecords.length === 0 && updateLocalRecords.length === 0) {
        return { records: [], counts: undefined };
    }

    onProgress({ phase: 'resolving', currentFile: 'Checking conflicts...' });

    const resolveItems: SyncConflictResolveItem[] = [];

    for (const conflict of conflictRecords) {
        const relativePath = conflict.filePath;

        try {
            const fullPath = ctx.decodedRepoPath ? `${ctx.decodedRepoPath}/${relativePath}` : relativePath;
            if (!fileOps.fileExists(fullPath)) {
                console.error('Conflict file not found locally:', relativePath);
                continue;
            }

            const fileContentBase64 = await fileOps.readFileBase64(fullPath);
            resolveItems.push({
                path: relativePath,
                songId: conflict.songId,
                fileContentBase64,
                localModifiedAt: safeToIsoString((conflict.data as any)?.localModifiedAt ? new Date((conflict.data as any).localModifiedAt) : new Date())!,
            });
        } catch (e) {
            console.error('Failed to read file for conflict resolution:', relativePath, e);
        }
    }

    const potentialUpdateItems: SyncPotentialUpdateResolveItem[] = [];

    for (const update of updateLocalRecords) {
        const relativePath = update.filePath;

        try {
            const fullPath = ctx.decodedRepoPath ? `${ctx.decodedRepoPath}/${relativePath}` : relativePath;
            if (!fileOps.fileExists(fullPath)) {
                console.error('Potential update file not found locally:', relativePath);
                continue;
            }

            const fileContentBase64 = await fileOps.readFileBase64(fullPath);
            const updateData = update.data as any;
            potentialUpdateItems.push({
                path: relativePath,
                songId: update.songId!,
                fileContentBase64,
                localModifiedAt: safeToIsoString(updateData?.localModifiedAt ? new Date(updateData.localModifiedAt) : new Date())!,
                lastSyncedAt: safeToIsoString(updateData?.lastSyncedAt ? new Date(updateData.lastSyncedAt) : new Date())!,
            });
        } catch (e) {
            console.error('Failed to read file for potential update resolution:', relativePath, e);
        }
    }

    if (resolveItems.length === 0 && potentialUpdateItems.length === 0) {
        return { records: [], counts: undefined };
    }

    try {
        const resolveResponse = await apiClient.resolveConflicts(ctx.deviceId, ctx.sessionId!, {
            conflicts: resolveItems,
            potentialUpdates: potentialUpdateItems,
        });

        ctx.result = addDeltaToResult(ctx.result, resolveResponse.counts ?? EMPTY_COUNTS);

        for (const record of resolveResponse.records) {
            switch (record.action) {
                case 'UpdateTimestamp':
                    console.log('Resolved conflict for', record.filePath, ':', record.reason);
                    break;
                case 'Conflict':
                    if (ctx.options.treatConflictsAsErrors) {
                        console.error('Conflict (as error):', record.filePath, record.reason);
                        ctx.result.error++;
                        ctx.result.conflict++;
                    } else {
                        const resolution = await userPrompt.promptConflictResolution(record.filePath);
                        if (resolution === 'upload') {
                            toUpdatePaths.add(record.filePath);
                        } else {
                            ctx.result.error++;
                        }
                    }
                    break;
                case 'CreateRemote':
                case 'UpdateRemote':
                    toUpdatePaths.add(record.filePath);
                    break;
                case 'UpdateLocal':
                case 'Rename':
                case 'Error':
                    console.log(`${record.action} action for ${record.filePath}: ${record.reason}`);
                    break;
            }
        }

        return { records: resolveResponse.records, counts: resolveResponse.counts };
    } catch (e) {
        console.error('Failed to resolve conflicts:', e);
        return { records: [], counts: undefined };
    }
}